### PR TITLE
fix: only provide the units belonging to the app in Relation.units

### LINKED ISSUE #1109: remote unit data is inaccessible in relation-departed
The [docs](https://juju.is/docs/sdk/relation-name-relation-departed-event#heading--relation-data) say:

> A unit’s relation settings persist beyond its own departure from the relation
[...]
> During a relation-departed hook, relation settings can still be read (with relation-get) [...] All units will still be able to see all other units

However, it's not possible (without using a private attribute) to access remote data from ops.

This [pair of example charms](https://gist.github.com/tonyandrewmeyer/695a91eff9eae093a073683739b42aef) demonstrate the issue:

* 'provider-charm' puts some example data into the local unit relation databag on `relation-joined` and does nothing else.
* 'require-charm' logs that data on `relation-changed` (just to show it's accessible normally) and attempts to log that data on `relation-departed`. However, that will fail, unless we get the data with manually creating a `RelationDataContent` object, which requires using `model._backend`.

To use the charms:

1. `charmcraft pack` each charm, and `juju deploy` them into a model (I used k8s but I believe this will be the case for machine as well)
2. `juju integrate` the two charms together
3. Wait for everything to settle down (you'll see that the data is in the relation databag from the `relation-changed` debug-log output)
4. Use `juju remove-relation` to remove the integration between the applications
5. You should see `departed: foo` and `departed (using private): foo` in `juju debug-log`, but instead will see `departed: COULD NOT GET IT.` and `departed (using private): foo`.

The reason for this is that `Relation.units` is populated by running Juju's `relation-list`, which doesn't have the departing relation in the `relation-departed` event. `Relation.data` is a `RelationData` object, and the `._data` attribute (which `__getitem__` exposes) is populated with (a) data for the local unit, (b) data for the local app, and (c) data for each unit in the relation's `.units`. In (e.g.) `relation-changed`, this means that the remote unit's data will be there, but in `relation-departed` it will not.

Juju's `relation-get` will still provide access to the data, and this is meant to be available (according to the docs linked above and also [the Juju team](https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$R8_gCGrfhjcqM6k_zIaVFcoiWX6sw02aSmoxUcVxsnM?via=ubuntu.com&via=matrix.org&via=fsfe.org)). However, the only method ops provides to use `relation-get` is via the `RelationDataContent` class (which needs the backend passed to instantiate an object) or `RelationData` (which needs the unit to be in `relation-list` as described above).

It seems like we should make this data available. It seems cleanest for it to be available in `Relation.data` like it would be outside of `relation-departed` - the departed unit is available in the environment and in the `event` object, so there would need to be a bit of work to get it into `Relation.data` without passing too much context around, but it's certainly feasible.

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
The previous PR for #1109 wrongly added the `REMOTE_UNIT` unit to all relations, but it should only be ones where the app matches (and in particular, is never for peer relations). This fixes that, and adds an extra unit test for verification.

To be extra sure, an integration test is added that verifies that the correct units are returned. The test/integration conftest was quite specific to tracing, so this also makes it more generalised.

I'd like to have a few integration tests for ops, and the door has been opened with tracing, but I'm happy to discuss this if people feel it's not worth doing for this case (I had to more-or-less make it anyway to verify the fix). It does form quite a significant portion of the changes here.

Fixes #1109

### CODE REVIEW

File: test/integration/test_relation.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    ingress_units = set(status.get_units(ingress))
+    task = juju.run('test-relation/0', 'get-units')
+    assert task.success, task.message
+    ops_units = json.loads(task.results['units'])
+
+    assert all(
+        unit.startswith(f'{app}/')
+        for rel, app in [('db', db), ('ingress', ingress), ('peer', 'test-relation')]
+        for unit in ops_units[rel]
+    )
```
Comment (james-garner-canonical) [inline test/integration/test_relation.py]: One minor suggestion
```suggestion
    assert all(unit.startswith('db/') for unit in ops_units[db])
    assert all(unit.startswith('ingress/') for unit in ops_units[ingress])
    assert all(unit.startswith('peer/') for unit in ops_units['test-relation])
```

File: ops/model.py [hunk 2]
```diff
@@ -1763,7 +1763,11 @@ def __init__(
 
         # In relation-departed `relation-list` doesn't include the remote unit,
         # but the data should still be available.
-        if _remote_unit is not None:
+        if (
+            _remote_unit is not None
+            and not is_peer
+            and _remote_unit.name.startswith(relation_name)
```
Comment (dimaqq) [inline ops/model.py]: This looks weird to me.

I would have imagined that unit name and relation names to be incomparable.

Can you provide examples for these strings?

What about the slashes, as in `"application/0".startswith("app")` but these are different apps, so it should be something like `"application/0".startswith("app/")`.

Comment (tonyandrewmeyer) [reply ops/model.py]: Ah, nice catch, thanks. Yes, the test was failing to catch this because the remote app name was set to the same name as the relation name. I've fixed the test and the code, and also handled the edge case.

File: ops/model.py [hunk 3]
```diff
@@ -1777,6 +1772,15 @@ def __init__(
         self.app = typing.cast('Application', app)
         self.data = RelationData(self, our_unit, backend)
 
+        # In relation-departed `relation-list` doesn't include the remote unit,
+        # but the data should still be available.
+        if (
+            _remote_unit is not None
+            and not is_peer
+            and _remote_unit.name.startswith(f'{self.app.name}/')
```
Comment (dimaqq) [inline ops/model.py:1780]: **curiosity:** is it possible to for an app to be related to itself over a non-peer relation anyway?

File: test/integration/conftest.py [hunk 4]
```diff
@@ -29,64 +29,90 @@
...
-            access_key='accesskey',
-            secret_key='mysoverysecretkey',
-            secure=False,
-        )
 
-        found = mc_client.bucket_exists('tempo')
-        if not found:
-            mc_client.make_bucket('tempo')
+@pytest.fixture
+def tracing_juju(juju: jubilant.Juju) -> Generator[jubilant.Juju]:
```
Comment (dimaqq) [inline test/integration/conftest.py:40]: **praise:** thank you for refactoring this file!

File: ops/model.py (hunk 3 above)
Comment (tonyandrewmeyer) [reply ops/model.py:1780]: No, I don't think so.

File: test/integration/conftest.py [hunk 5]
```diff
@@ -29,64 +29,90 @@
...
-    """Prepare and return the test charm directory.
+def tracing_charm_dir(pytestconfig: pytest.Config) -> Generator[pathlib.Path]:
+    """Prepare and return the test_tracing charm directory.
 
     Builds and injects `ops` and `ops-tracing` from the local checkout in to the
     charm's dependencies. Cleans up afterwards.
     """
-    charm_dir = pytestconfig.rootpath / 'test/charms/test_tracing'
+    charm_dir = pytestconfig.rootpath / 'test/charms/test_tracing'  # type: ignore
+    assert isinstance(charm_dir, pathlib.Path)
```
Comment (dimaqq) [inline test/integration/conftest.py:95]: Weirdly I don't need `type: ignore` or `isinstance` check here.

If it's really needed, why not flip this like so?

```suggestion
    assert isinstance(pytestconfig.rootpath, pathlib.Path)
    charm_dir = pytestconfig.rootpath / 'test/charms/test_tracing'
```

Comment (tonyandrewmeyer) [reply test/integration/conftest.py:95]: Maybe this is because of a newer pyright? This was done 3 months ago and I think we bumped the version since then.

No objections to opening a new PR that makes improvements here.
